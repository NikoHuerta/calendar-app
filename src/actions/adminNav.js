import { types } from "../types/types";

export const setNavigation = ( page ) => ({
    type: types.adminNavigation,
    payload: page
});