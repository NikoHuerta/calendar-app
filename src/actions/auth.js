import Swal from "sweetalert2";

import { fetchSinToken, fetchAxios } from "../helpers/fetch"
import { types } from "../types/types";


export const startLogin = (email, password) => {
    return async ( dispatch ) => {

        // const resp = await fetchSinToken('auth', { email, password }, 'POST');
        // const body = await resp.json();
        const resp = await fetchAxios('auth', { email, password }, 'POST');
        const { data: body } = resp;

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({ 
                    uid: body.uid,
                    name: body.name
                }) );
        } else {

            const errorMsg = body.errors ? Object.values( body.errors )[0].msg : body.msg;
            Swal.fire('Error', errorMsg, 'error');
        }
    }
}

export const startRegister = (email, password, name) => {
    return async (dispatch) => {

        // const resp = await fetchSinToken('usuarios', { name, email, password, rol: 'USER_ROLE' }, 'POST');
        // const body = await resp.json();
        const resp = await fetchAxios('usuarios', { name, email, password, rol: 'USER_ROLE' }, 'POST');
        const { data: body } = resp;

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({ 
                uid: body.uid,
                name: body.name
            }) );
            
        } else {
            
            const errorMsg = body.errors ? Object.values( body.errors )[0].msg : body.msg;
            Swal.fire('Error', errorMsg, 'error');
        }
    }
}

export const startGoogleLogin = (idToken) => {
    return async (dispatch) => {

        // const resp = await fetchSinToken('auth/google', { id_token: idToken }, 'POST');
        // const body = await resp.json();
        const resp = await fetchAxios('auth/google', { id_token: idToken }, 'POST');
        const { data: body } = resp;

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({ 
                uid: body.usuario.uid,
                name: body.usuario.name
            }) );

        } else {
            
            const errorMsg = body.errors ? Object.values( body.errors )[0].msg : body.msg;
            Swal.fire('Error', errorMsg, 'error');
        }
    }
}

export const startFacebookLogin = (idToken, name, email) => {
    return async (dispatch) => {
        // const resp = await fetchSinToken('auth/facebook', { id_token: idToken, name, email }, 'POST');
        // const body = await resp.json();

        const resp = await fetchAxios('auth/facebook', { id_token: idToken, name, email }, 'POST');
        const { data: body } = resp;


        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({ 
                uid: body.usuario.uid,
                name: body.usuario.name
            }) );

        } else {
            
            const errorMsg = body.errors ? Object.values( body.errors )[0].msg : body.msg;
            Swal.fire('Error', errorMsg, 'error');
        }
    }
}

export const startChecking = () => {
    return async (dispatch) => {
        const resp = await fetchAxios('auth/renew', undefined,'GET',undefined, localStorage.getItem('token'));
        const { data: body } = resp;

        if(body.ok){

            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({ 
                uid: body.uid,
                name: body.name
            }) );

        } else {
            dispatch( checkingFinish() );
        }

    }
}

const checkingFinish = () => ({ type: types.authCheckingFinish });

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});

const logout = () => ({ type: types.authLogout });

export const startLogout = () => {
    return( dispatch ) => {
        localStorage.clear();
        dispatch( logout() );
    }
}
