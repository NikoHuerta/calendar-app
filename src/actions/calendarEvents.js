import { types } from "../types/types"

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
})

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventRemoveActive = () => ({ type: types.eventRemoveActive });

export const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
}); 

export const eventDeleted = () => ({ type: types.eventDeleted });

export const eventSelectedDate = (dates) => ({ 
    type: types.eventSelectedDate,
    payload: dates
});

export const eventRemoveSelectedDate = () => ({ type: types.eventRemoveSelectedDate });