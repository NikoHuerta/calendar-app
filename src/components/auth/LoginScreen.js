import React from 'react';
import './login.css';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import validator from 'validator';

import GoogleLogin from 'react-google-login';
import FacebookLogin from 'react-facebook-login/dist/facebook-login-render-props';
import { GoogleLoginButton, FacebookLoginButton } from "react-social-login-buttons";

import { useForm } from '../../hooks/useForm';
import { startFacebookLogin, startGoogleLogin, startLogin, startRegister } from '../../actions/auth';


export const LoginScreen = () => {

    const dispatch = useDispatch();
    const [formLoginValues, handleLoginInputChange] = useForm({
        lEmail: 'nhuertaf@icloud.com',
        lPassword: '123456'
    });

    const [formRegisterValues, handleRegisterInputChange] = useForm({
        rName: 'Nicolas',
        rEmail: 'nicolas@test.com',
        rPassword1: '123456',
        rPassword2: '123456'
    });

    const { rName, rEmail, rPassword1, rPassword2 } = formRegisterValues;
    const { lEmail, lPassword } = formLoginValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( startLogin( lEmail, lPassword ) );
    }

    const handleRegister = (e) => {
        e.preventDefault();

        if( rPassword1 !== rPassword2 )
            return Swal.fire('Error', 'Las contraseñas tienen que ser iguales', 'error');
        if( !validator.isEmail( rEmail ))
            return Swal.fire('Error', 'El correo es inválido', 'error')
        if( !validator.isLength( rName, { min:6, max: undefined }) )
            return Swal.fire('Error', 'El nombre debe ser de mas de 6 caracteres', 'error');
        
        dispatch( startRegister(rEmail, rPassword1, rName) );
    }

    const handleGoogleLogin = ( response ) => {
        const { tokenId } = response;
        dispatch( startGoogleLogin(tokenId) );
    };

    const handleFacebookLogin = ( response ) => {
        const { accessToken, name, email } = response;
        dispatch( startFacebookLogin(accessToken, name, email) );
    };


    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name="lEmail"
                                value={ lEmail }
                                onChange={ handleLoginInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name="lPassword"
                                value={ lPassword }
                                onChange={ handleLoginInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name = "rName"
                                value = { rName }
                                onChange = { handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name = "rEmail"
                                value = { rEmail }
                                onChange = { handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña" 
                                name = "rPassword1"
                                value = { rPassword1 }
                                onChange = { handleRegisterInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña" 
                                name = "rPassword2"
                                value = { rPassword2 }
                                onChange = { handleRegisterInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
                
                <div className="auth__social-networks">
                    <h3>Login with social networks</h3>
                    <div className="col-md-6">
                        <GoogleLogin 
                            clientId={ process.env.REACT_APP_GOOGLE_CLIENT }
                            buttonText="Login with Google"
                            onSuccess={ handleGoogleLogin }
                            onFailure={ handleGoogleLogin }
                            cookiePolicy={'single_host_origin'}
                            isSignedIn={ false }
                            render={renderProps => (
                                <GoogleLoginButton onClick={renderProps.onClick} ></GoogleLoginButton>
                                )}
                        />
                    </div>

                    <div className="col-md-6">
                        <FacebookLogin 
                            appId={ process.env.REACT_APP_FACEBOOK_CLIENT }
                            fields="name,email,picture"
                            callback={ handleFacebookLogin }
                            autoLoad={ false }
                            render={renderProps => (
                                <FacebookLoginButton onClick={renderProps.onClick} ></FacebookLoginButton>
                                )}
                        />
                    </div>
                </div>
                
            </div>
        </div>
    )
}