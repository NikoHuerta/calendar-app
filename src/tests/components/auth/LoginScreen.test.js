import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import Swal from 'sweetalert2';




import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', ()=> ({
    fire: jest.fn()
}));


const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {
    calendar: {
        activeEvent: '12345678'
    }
};
let store = mockStore( initState );
store.dispatch = jest.fn(); //solo importa que se dispare, por eso el mock


const wrapper = mount(
    <Provider store={ store }>
        <LoginScreen />
    </Provider>
);

describe('Pruebas en <LoginScreen />', () => { 

    beforeEach( () => {
        jest.clearAllMocks();
    });

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar el dispatch del login', () => {
        
        wrapper.find('input[name="lEmail"]').simulate('change', {
            target: {
                name: 'lEmail',
                value: 'nico@nico.nico'
            }
        });
        wrapper.find('input[name="lPassword"]').simulate('change', {
            target: {
                name: 'lPassword',
                value: '123456'
            }
        });

        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault(){}
        });

        expect(startLogin).toHaveBeenCalledWith('nico@nico.nico','123456');
    });

    test('Debe de llamar el dispatch del register', () => {
        
        wrapper.find('input[name="rName"]').simulate('change', {
            target: {
                name: 'rName',
                value: 'nicolas huerta'
            }
        });
        wrapper.find('input[name="rEmail"]').simulate('change', {
            target: {
                name: 'rEmail',
                value: 'nico@nico.cl'
            }
        });
        wrapper.find('input[name="rPassword1"]').simulate('change', {
            target: {
                name: 'rPassword1',
                value: '123456'
            }
        });
        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: '123456'
            }
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault(){}
        });
        expect(Swal.fire).not.toHaveBeenCalled();
        expect(startRegister).toHaveBeenCalledWith('nico@nico.cl','123456', 'nicolas huerta');
    });

    test('No hay registro si las contraseñas son diferentes', () => {
        wrapper.find('input[name="rName"]').simulate('change', {
            target: {
                name: 'rName',
                value: 'nicolas huerta'
            }
        });
        wrapper.find('input[name="rEmail"]').simulate('change', {
            target: {
                name: 'rEmail',
                value: 'nico@nico.cl'
            }
        });
        wrapper.find('input[name="rPassword1"]').simulate('change', {
            target: {
                name: 'rPassword1',
                value: '123456789'
            }
        });
        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: '1234567'
            }
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault(){}
        });

        // expect(startRegister).toHaveBeenCalledTimes(0);
        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseñas tienen que ser iguales', 'error');
    });

});