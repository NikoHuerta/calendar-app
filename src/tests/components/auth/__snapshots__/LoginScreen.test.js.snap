// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe de mostrarse correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="loginEmail"
          >
            Email Address
          </label>
          <input
            className="form-control"
            id="loginEmail"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="nhuertaf@icloud.com"
          />
        </div>
        <div
          className="form-group mt-3"
        >
          <label
            className="form-label"
            htmlFor="loginPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            id="loginPassword"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group mt-4 text-center"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2 text-light"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="registerName"
          >
            Nombre
          </label>
          <input
            className="form-control"
            id="registerName"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Nicolas"
          />
        </div>
        <div
          className="form-group mt-3"
        >
          <label
            className="form-label"
            htmlFor="registerEmail"
          >
            Email Address
          </label>
          <input
            className="form-control"
            id="registerEmail"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="nicolas@test.com"
          />
        </div>
        <div
          className="form-group mt-3"
        >
          <label
            className="form-label"
            htmlFor="registerPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            id="registerPassword"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group mt-3"
        >
          <label
            className="form-label"
            htmlFor="registerPassword2"
          >
             Confirmar password
          </label>
          <input
            className="form-control"
            id="registerPassword2"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group mt-4 text-center"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
